File: daedalos-tools/gates/lib/checker.py
"""Gate checking and approval flow."""

import os
import sys
import json
import time
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Optional, Dict, Any, List

try:
    from .config import (
        SupervisionConfig,
        load_project_config,
        GATE_ACTIONS,
    )
except ImportError:
    from config import (
        SupervisionConfig,
        load_project_config,
        GATE_ACTIONS,
    )


# Data directory for gate history
DATA_DIR = Path(os.environ.get("XDG_DATA_HOME", Path.home() / ".local" / "share")) / "daedalos" / "gates"


@dataclass
class GateRequest:
    """A request to pass through a gate."""

    gate: str
    action: str
    context: Dict[str, Any]
    timestamp: float
    source: str  # Which tool is requesting


@dataclass
class GateResult:
    """Result of a gate check."""

    allowed: bool
    action: str  # What action was taken (allow, notify, approve, deny)
    reason: str
    approved_by: Optional[str] = None  # "auto", "user", "timeout"


def ensure_data_dir():
    """Ensure data directory exists."""
    DATA_DIR.mkdir(parents=True, exist_ok=True)


def log_gate_event(request: GateRequest, result: GateResult) -> None:
    """Log a gate event for history/audit."""
    ensure_data_dir()

    event = {
        "timestamp": request.timestamp,
        "gate": request.gate,
        "action": request.action,
        "context": request.context,
        "source": request.source,
        "result": {
            "allowed": result.allowed,
            "action": result.action,
            "reason": result.reason,
            "approved_by": result.approved_by,
        }
    }

    # Append to daily log file
    date_str = datetime.fromtimestamp(request.timestamp).strftime("%Y-%m-%d")
    log_file = DATA_DIR / f"gates-{date_str}.jsonl"

    with open(log_file, "a") as f:
        f.write(json.dumps(event) + "\n")


def check_gate(
    gate: str,
    context: Optional[Dict[str, Any]] = None,
    source: str = "unknown",
    config: Optional[SupervisionConfig] = None,
    interactive: bool = True,
) -> GateResult:
    """
    Check if an action is allowed through a gate.

    Args:
        gate: The gate name (e.g., "file_delete", "git_push")
        context: Additional context about the action
        source: Which tool is making the request
        config: Supervision config (loaded if not provided)
        interactive: Whether to prompt for approval if needed

    Returns:
        GateResult indicating if the action is allowed
    """
    if config is None:
        config = load_project_config()

    if context is None:
        context = {}

    # Check for sensitive path override
    if "path" in context and config.is_sensitive_path(context["path"]):
        gate_action = config.get_gate("sensitive_file")
    else:
        gate_action = config.get_gate(gate)

    request = GateRequest(
        gate=gate,
        action=gate_action,
        context=context,
        timestamp=time.time(),
        source=source,
    )

    if gate_action == "allow":
        result = GateResult(
            allowed=True,
            action="allow",
            reason="Gate configured to allow",
            approved_by="auto",
        )

    elif gate_action == "notify":
        # Notify but don't block
        notify_user(request)
        result = GateResult(
            allowed=True,
            action="notify",
            reason="User notified, proceeding",
            approved_by="auto",
        )

    elif gate_action == "deny":
        result = GateResult(
            allowed=False,
            action="deny",
            reason="Gate configured to deny",
            approved_by="auto",
        )

    elif gate_action == "approve":
        if interactive:
            approved = prompt_for_approval(request)
            result = GateResult(
                allowed=approved,
                action="approve",
                reason="User approved" if approved else "User denied",
                approved_by="user" if approved else None,
            )
        else:
            # Non-interactive mode - check for pre-approval or deny
            result = GateResult(
                allowed=False,
                action="approve",
                reason="Approval required but running non-interactively",
                approved_by=None,
            )

    else:
        # Unknown action, default to approve
        result = GateResult(
            allowed=False,
            action="approve",
            reason=f"Unknown gate action: {gate_action}",
            approved_by=None,
        )

    # Log the event
    log_gate_event(request, result)

    return result


def notify_user(request: GateRequest) -> None:
    """Send a notification to the user (non-blocking)."""
    # Format the notification
    msg = format_gate_message(request, is_notification=True)

    # Print to stderr so it's visible but doesn't interfere with stdout
    print(f"\033[33m[gates]\033[0m {msg}", file=sys.stderr)

    # Could also send to system notification daemon, tmux, etc.
    # For now, just print


def prompt_for_approval(request: GateRequest) -> bool:
    """Prompt user for approval (blocking)."""
    msg = format_gate_message(request, is_notification=False)

    # Check if we're in a terminal
    if not sys.stdin.isatty():
        print(f"\033[31m[gates]\033[0m Approval needed but no TTY: {msg}", file=sys.stderr)
        return False

    print(f"\n\033[33m[gates]\033[0m {msg}", file=sys.stderr)
    print(f"\033[33m[gates]\033[0m Allow this action? [y/N] ", file=sys.stderr, end="")

    try:
        response = input().strip().lower()
        return response in ("y", "yes")
    except (EOFError, KeyboardInterrupt):
        print("", file=sys.stderr)
        return False


def format_gate_message(request: GateRequest, is_notification: bool = False) -> str:
    """Format a gate request for display."""
    parts = [f"Gate: {request.gate}"]

    if request.source != "unknown":
        parts.append(f"Source: {request.source}")

    # Add relevant context
    ctx = request.context
    if "path" in ctx:
        parts.append(f"Path: {ctx['path']}")
    if "command" in ctx:
        cmd = ctx["command"]
        if len(cmd) > 60:
            cmd = cmd[:57] + "..."
        parts.append(f"Command: {cmd}")
    if "description" in ctx:
        parts.append(f"Action: {ctx['description']}")

    prefix = "Notification" if is_notification else "Approval required"
    return f"{prefix} - " + " | ".join(parts)


def get_gate_history(
    gate: Optional[str] = None,
    days: int = 7,
    limit: int = 100,
) -> List[Dict[str, Any]]:
    """Get gate check history."""
    ensure_data_dir()

    events = []
    now = datetime.now()

    for i in range(days):
        date = now.replace(hour=0, minute=0, second=0, microsecond=0)
        date_str = date.strftime("%Y-%m-%d")
        log_file = DATA_DIR / f"gates-{date_str}.jsonl"

        if log_file.exists():
            with open(log_file) as f:
                for line in f:
                    try:
                        event = json.loads(line)
                        if gate is None or event.get("gate") == gate:
                            events.append(event)
                    except json.JSONDecodeError:
                        pass

        now = now.replace(day=now.day - 1)

    # Sort by timestamp descending and limit
    events.sort(key=lambda x: x.get("timestamp", 0), reverse=True)
    return events[:limit]


def check_autonomy_limits(
    config: SupervisionConfig,
    iterations: int = 0,
    file_changes: int = 0,
    lines_changed: int = 0,
) -> Optional[str]:
    """
    Check if we've exceeded autonomy limits.

    Returns a reason string if limits exceeded, None otherwise.
    """
    limits = config.autonomy

    if iterations > limits.get("max_iterations", float("inf")):
        return f"Exceeded max iterations ({iterations}/{limits['max_iterations']})"

    if file_changes > limits.get("max_file_changes", float("inf")):
        return f"Exceeded max file changes ({file_changes}/{limits['max_file_changes']})"

    if lines_changed > limits.get("max_lines_changed", float("inf")):
        return f"Exceeded max lines changed ({lines_changed}/{limits['max_lines_changed']})"

    return None

File: crates/gates/src/checker.rs
//! Gate checking and history logging
//!
//! Handles gate checks, approval flows, and history recording.

use anyhow::{Context, Result};
use chrono::{DateTime, Local, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs::{self, File, OpenOptions};
use std::io::{BufRead, BufReader, Write};
use std::path::PathBuf;

use crate::config::{load_project_config, GateAction, SupervisionConfig};

/// Get the data directory for gate history
pub fn get_data_dir() -> PathBuf {
    dirs::data_dir()
        .unwrap_or_else(|| PathBuf::from("~/.local/share"))
        .join("daedalos")
        .join("gates")
}

/// A request to pass through a gate
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GateRequest {
    pub gate: String,
    pub action: String,
    pub context: HashMap<String, serde_json::Value>,
    pub timestamp: f64,
    pub source: String,
}

/// Result of a gate check
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GateResult {
    pub allowed: bool,
    pub action: String,
    pub reason: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approved_by: Option<String>,
}

/// Event logged to history
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GateEvent {
    pub timestamp: f64,
    pub gate: String,
    pub action: String,
    pub context: HashMap<String, serde_json::Value>,
    pub source: String,
    pub result: GateEventResult,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GateEventResult {
    pub allowed: bool,
    pub action: String,
    pub reason: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approved_by: Option<String>,
}

impl From<&GateResult> for GateEventResult {
    fn from(result: &GateResult) -> Self {
        Self {
            allowed: result.allowed,
            action: result.action.clone(),
            reason: result.reason.clone(),
            approved_by: result.approved_by.clone(),
        }
    }
}

/// Ensure data directory exists
fn ensure_data_dir() -> Result<PathBuf> {
    let data_dir = get_data_dir();
    fs::create_dir_all(&data_dir)
        .with_context(|| format!("Failed to create data directory: {}", data_dir.display()))?;
    Ok(data_dir)
}

/// Log a gate event for history/audit
fn log_gate_event(request: &GateRequest, result: &GateResult) -> Result<()> {
    let data_dir = ensure_data_dir()?;

    let event = GateEvent {
        timestamp: request.timestamp,
        gate: request.gate.clone(),
        action: request.action.clone(),
        context: request.context.clone(),
        source: request.source.clone(),
        result: result.into(),
    };

    // Append to daily log file
    let timestamp = request.timestamp as i64;
    let datetime = DateTime::from_timestamp(timestamp, 0)
        .unwrap_or_else(|| Utc::now());
    let date_str = datetime.format("%Y-%m-%d").to_string();
    let log_file = data_dir.join(format!("gates-{}.jsonl", date_str));

    let mut file = OpenOptions::new()
        .create(true)
        .append(true)
        .open(&log_file)
        .with_context(|| format!("Failed to open log file: {}", log_file.display()))?;

    let json = serde_json::to_string(&event).context("Failed to serialize event")?;
    writeln!(file, "{}", json).context("Failed to write to log file")?;

    Ok(())
}

/// Check if an action is allowed through a gate
pub fn check_gate(
    gate: &str,
    context: Option<HashMap<String, serde_json::Value>>,
    source: &str,
    config: Option<&SupervisionConfig>,
) -> Result<GateResult> {
    let config = match config {
        Some(c) => c.clone(),
        None => load_project_config(None)?,
    };

    let context = context.unwrap_or_default();
    let timestamp = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_secs_f64();

    // Check for sensitive path override
    let gate_action = if let Some(path) = context.get("path").and_then(|v| v.as_str()) {
        if config.is_sensitive_path(path) {
            config.get_gate("sensitive_file")
        } else {
            config.get_gate(gate)
        }
    } else {
        config.get_gate(gate)
    };

    let request = GateRequest {
        gate: gate.to_string(),
        action: gate_action.as_str().to_string(),
        context,
        timestamp,
        source: source.to_string(),
    };

    let result = match gate_action {
        GateAction::Allow => GateResult {
            allowed: true,
            action: "allow".into(),
            reason: "Gate configured to allow".into(),
            approved_by: Some("auto".into()),
        },
        GateAction::Notify => {
            // In non-interactive mode, we just log and proceed
            GateResult {
                allowed: true,
                action: "notify".into(),
                reason: "User notified, proceeding".into(),
                approved_by: Some("auto".into()),
            }
        }
        GateAction::Deny => GateResult {
            allowed: false,
            action: "deny".into(),
            reason: "Gate configured to deny".into(),
            approved_by: Some("auto".into()),
        },
        GateAction::Approve => {
            // Non-interactive mode - approval required but we can't prompt
            GateResult {
                allowed: false,
                action: "approve".into(),
                reason: "Approval required but running non-interactively".into(),
                approved_by: None,
            }
        }
    };

    // Log the event
    if let Err(e) = log_gate_event(&request, &result) {
        eprintln!("Warning: Failed to log gate event: {}", e);
    }

    Ok(result)
}

/// Get gate check history
pub fn get_gate_history(
    gate: Option<&str>,
    days: u32,
    limit: usize,
) -> Result<Vec<GateEvent>> {
    let data_dir = ensure_data_dir()?;
    let mut events = Vec::new();

    let today = Local::now().date_naive();

    // Search from tomorrow (UTC may be ahead of local) back to days ago
    for i in -1..days as i64 {
        let date = today - chrono::Duration::days(i);
        let date_str = date.format("%Y-%m-%d").to_string();
        let log_file = data_dir.join(format!("gates-{}.jsonl", date_str));

        if log_file.exists() {
            let file = File::open(&log_file)?;
            let reader = BufReader::new(file);

            for line in reader.lines() {
                if let Ok(line) = line {
                    if let Ok(event) = serde_json::from_str::<GateEvent>(&line) {
                        if gate.is_none() || gate == Some(event.gate.as_str()) {
                            events.push(event);
                        }
                    }
                }
            }
        }
    }

    // Sort by timestamp descending
    events.sort_by(|a, b| b.timestamp.partial_cmp(&a.timestamp).unwrap_or(std::cmp::Ordering::Equal));

    // Apply limit
    events.truncate(limit);

    Ok(events)
}

/// Check autonomy limits
#[allow(dead_code)]
pub fn check_autonomy_limits(
    config: &SupervisionConfig,
    iterations: u32,
    file_changes: u32,
    lines_changed: u32,
) -> Option<String> {
    if iterations > config.autonomy.max_iterations {
        return Some(format!(
            "Exceeded max iterations ({}/{})",
            iterations, config.autonomy.max_iterations
        ));
    }

    if file_changes > config.autonomy.max_file_changes {
        return Some(format!(
            "Exceeded max file changes ({}/{})",
            file_changes, config.autonomy.max_file_changes
        ));
    }

    if lines_changed > config.autonomy.max_lines_changed {
        return Some(format!(
            "Exceeded max lines changed ({}/{})",
            lines_changed, config.autonomy.max_lines_changed
        ));
    }

    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_gate_result_serialization() {
        let result = GateResult {
            allowed: true,
            action: "allow".into(),
            reason: "Test".into(),
            approved_by: Some("auto".into()),
        };
        let json = serde_json::to_string(&result).unwrap();
        assert!(json.contains("allowed"));
    }

    #[test]
    fn test_check_autonomy_limits() {
        let config = SupervisionConfig::default();

        // Under limits
        assert!(check_autonomy_limits(&config, 10, 50, 500).is_none());

        // Over iterations limit
        let result = check_autonomy_limits(&config, 100, 50, 500);
        assert!(result.is_some());
        assert!(result.unwrap().contains("iterations"));
    }
}
